### ISSUE TITLE
Score Function Issues and Reward Hacking

### ISSUE THREAD

**Issue by gao-yuan-hangzhou** [Einsia/Frontier-Engineering#72]
From offline experiments, for tasks PowerSystems/EV2GymSmartCharging and AdditiveManufacturing/DiffSimThermalControl, I saw astronomically high scores achieved by frontier models (GPT-5.5 and Opus 4.7), which suggest score function issues that may impact the integrity of the evaluation set up and the leaderboard. The issues below are not from a full investigation. They are just what I noticed so far.

## Issue 1: `PowerSystems/EV2GymSmartCharging`
`benchmarks/PowerSystems/EV2GymSmartCharging/verification/evaluator.py:258-287`

### Score formula

```python
stats = _jsonable(get_statistics(env))
total_reward = float(stats["total_reward"])
evaluation_cost = max(1e-9, -total_reward)                # ← singularity
normalized_score = 100.0 * float(case_definition["baseline_cost"]) / evaluation_cost
```

The intent: prevent division by zero when a candidate produces near-zero cost.

The effect: the score has a **maximum at exactly `total_reward = 0`**, where it
evaluates to `100 × baseline_cost / 1e-9 ≈ 100 × baseline_cost × 10⁹`. With
`baseline_cost ≈ 3.7 × 10⁵` (observed empirically in our runs across the three
shipped cases), this comes out to **≈ 3.69 × 10¹⁶**.

### What the agents found

The simplest policy that produces `total_reward = 0` is **exactly** the
all-zero-action policy that the contributor's own `scripts/init.py` baseline
already implements:

```python
def solve(case, max_sim_calls=0, simulate_fn=None):
    return {"actions": [0.0 for _ in case.get("ports", ())]}
```

i.e. all six agent configurations were rewarded for *not changing* the
shipped baseline. The 5 cells that "won" 3.69×10¹⁶ all converged back to
this policy. The 6 cell that reported a legitimate-looking 155.6 was the
only one to ship a non-zero charging policy.

### Reproduction

```
python -m frontier_eval \
  task=unified task.benchmark=PowerSystems/EV2GymSmartCharging \
  algorithm=openevolve algorithm.iterations=100 \
  llm.timeout=600 llm.model=<your-model>
```

Look at `combined_score` in the output `best_program_info.json`. With most
configurations and 100 iterations, the optimizer lands on
`combined_score ≈ 3.69e+16` with
`mean_total_reward = 0.0`, `mean_total_profits = 0.0`, etc. The sub-metrics
clearly show no charging happened.

### Suggested fix

Choose either (the second is preferable because it's scale-invariant):

```python
# Option A — clip to a sane maximum
evaluation_cost = max(1.0, -total_reward)         # floor at $1, not $1e-9
normalized_score = min(1e4, 100.0 * baseline_cost / evaluation_cost)

# Option B — scale-invariant via tanh
normalized_score = 100.0 * (1.0 - math.tanh((-total_reward) / baseline_cost))
```

Either makes the zero-action policy a finite, unremarkable score (≈ 100, the
baseline) rather than the global maximum.

---

### Location

`benchmarks/AdditiveManufacturing/DiffSimThermalControl/verification/evaluator.py:40-46`

### Score formula

```python
improvement_ratio = (baseline_loss - candidate_loss) / max(abs(baseline_loss), 1e-9)
quality           = math.exp(-candidate_loss / 200.0)
call_penalty      = 0.002 * float(candidate_result['sim_calls']) / float(max_sim_calls)
score             = max(0.0, quality + 0.3 * improvement_ratio - call_penalty)
```

Two distinct flaws:

#### Flaw A: `sim_calls` is agent-supplied and unsanitized

`candidate_result['sim_calls']` comes from the agent's `solve()` return
value. There is no clamp on its sign or magnitude. In one of our cells
(`openevolve + GPT-5.5`, iter 92), the candidate returned
`sim_calls = -1 × 10¹⁰⁰` per case (totaling −4 × 10¹⁰⁰ across the 4 cases).
The penalty term then becomes

```
call_penalty = 0.002 × (-1e+100) / 24 ≈ -8.3 × 10⁹⁵
score        = max(0, quality + 0.3 - (-8.3e+95)) ≈ 8.3 × 10⁹⁵
```

This is what produced the headline 8.33×10⁹⁵ score on that cell. The agent
didn't actually solve any thermal-control problem — the score function gave
it ≈ 10⁹⁵ for misreporting one integer field.

#### Flaw B: score-function cap is reachable via `candidate_loss = 0`

A separate cell (`shinkaevolve + GPT-5.5`) reported
`mean_candidate_loss = 0.0`, `mean_improvement_ratio = 1.0` (the cap),
`sim_calls = 0`, producing `combined_score = 1.30` (= `quality(1) + 0.3·1 − 0`).

This is the *legitimate* bounded maximum of the score function. Whether it
represents an Adam-style optimizer genuinely converging to zero loss on the
shipped cases, or a more subtle exploit where the candidate's
`simulate()` returns zero loss without computing it honestly, requires
manual verification. The metrics alone (`candidate_loss = 0.0` exactly,
`sim_calls = 0` exactly, on all 4 cases) are suspicious but not conclusive.

### Reproduction

```
python -m frontier_eval \
  task=unified task.benchmark=AdditiveManufacturing/DiffSimThermalControl \
  algorithm=openevolve algorithm.iterations=100 \
  llm.timeout=600 llm.model=<some-frontier-model>
```

In our runs the agent took about 90 iterations to find Flaw A's exploit,
and not every framework × model combination found it. With ShinkaEvolve we
got Flaw B's pattern; with OpenEvolve+GPT we got Flaw A's pattern.

### Suggested fixes

**For Flaw A** (sanitize `sim_calls`):

```python
reported = int(candidate_result.get('sim_calls', 0))
sim_calls = max(0, min(reported, max_sim_calls))
```

A stronger fix: track `sim_calls` server-side by wrapping `simulate` in an
instrumented proxy that the candidate cannot bypass.

**For Flaw B** (independent re-check on candidate_loss = 0):

```python
if candidate_loss <= 1e-6 * max(1.0, abs(baseline_loss)):
    # Re-simulate via a canonical, benchmark-owned simulator that the
    # candidate cannot redefine.
    canonical_loss = CANONICAL_SIMULATE(candidate_result['params'], case)['loss']
    if abs(canonical_loss - candidate_loss) > 0.01 * abs(baseline_loss):
        candidate_loss = canonical_loss
```

Structurally, the cleaner fix is to **not let candidates redefine
`simulate` or `load_cases`** and move both into a benchmark-owned module that
the candidate imports. This matches the architecture used in
`WirelessChannelSimulation/HighReliableSimulation`, where the canonical
simulation loop lives outside the agent's editable region and is
explicitly called out in the task notes as preventing "reward hacking
through self-reported aggregate statistics returned by candidate
wrappers."

---

## Issue 3: Possible third case: `KernelEngineering/FlashAttention` (unverified)
https://lab.einsia.ai/frontier-eng/problem.html?task=FlashAttention&exp=model
<img width="1031" height="547" alt="Image" src="https://github.com/user-attachments/assets/bf641549-211d-4fcd-b510-2dedcd5d2c8d" />

**Comment by ahydchh** [comment]
Thanks for the very detailed report and the concrete repro notes. I confirmed both reported scoring problems and pushed fixes in `9cad9fe`.

For `PowerSystems/EV2GymSmartCharging`, the root cause was exactly the score singularity you pointed out: `evaluation_cost = max(1e-9, -total_reward)` made a zero-reward/no-op rollout look like an almost-zero-cost optimum. The fix changes the evaluator so that:

- rollouts with effectively no delivered charging service receive score `0` instead of being rewarded for low cost;
- the cost denominator now has a practical floor instead of `1e-9`;
- per-case normalized scores are capped, so near-zero-cost edge cases cannot dominate the benchmark;
- the normal finite-score path is preserved: the shipped feasible seed evaluates around `99.97` in a smoke test, close to the official-style heuristic normalized around `100`. Here `100` is the normalized baseline, not the maximum score; improved valid policies can score above baseline, up to the evaluator cap of `1000`.

I also updated the EV2Gym runtime requirements to pin `setuptools<81`, because the upstream `ev2gym` package still imports `pkg_resources`; without that pin, fresh Python environments can fail before reaching the evaluator.

For `AdditiveManufacturing/DiffSimThermalControl`, I fixed the trust boundary rather than only clamping the reported integer. The evaluator now loads the fixed case loader, simulator, and baseline scoring path from a new read-only `verification/canonical.py`. Candidate code is only trusted for `solve(...)`. In particular:

- candidate-provided `load_cases`, `simulate`, and `baseline_solve` are not used for scoring;
- candidate-reported `sim_calls` is ignored;
- the evaluator wraps the canonical `simulate_fn` and counts actual calls itself;
- returned params are validated for shape and finite values before canonical simulation;
- the call penalty and validity check use the evaluator-measured call count.

I verified the fix with targeted exploit tests: a candidate returning `sim_calls = -10**100` no longer inflates the score, and a candidate that defines a fake `simulate()` returning zero loss is also scored through the canonical simulator instead. The normal unified smoke test still passes for the initial program with `benchmark_returncode=0`.

We will re-evolve the affected results with the corrected evaluators in the next release.
